File: lib/extraction/align-evidence.ts
import type { ExtractedField, ExtractedInvoice, FieldStatus, InvoiceLineItem } from "@/lib/types";
import type { RawExtraction, RawField } from "@/lib/extraction/schema";
import type { TextLayerLine } from "@/lib/extraction/pdf-text-layer";

type Haystack = {
  text: string;
  page: number;
  box: [number, number, number, number];
};

/**
 * Two search tiers: individual text-layer lines, and consecutive same-page
 * pairs concatenated. Pairs exist only to recover a quote that fell across a
 * PDF line-wrap boundary (see pdf-generate.ts's wrapText) — they must stay a
 * strict fallback, searched only when no single line matches. Searching both
 * tiers together is wrong: any quote that is a prefix of line N will also
 * match the (N, N+1) pair, producing two different boxes for one true match
 * and a false "conflicting" status.
 */
function buildHaystacks(textLayer: TextLayerLine[]): { singles: Haystack[]; pairs: Haystack[] } {
  const singles: Haystack[] = textLayer.map((l) => ({ text: l.text, page: l.page, box: l.box }));
  const pairs: Haystack[] = [];

  for (let i = 0; i < textLayer.length - 1; i++) {
    const a = textLayer[i];
    const b = textLayer[i + 1];
    if (a.page !== b.page) continue;
    pairs.push({
      text: `${a.text} ${b.text}`,
      page: a.page,
      box: unionBox(a.box, b.box),
    });
  }

  return { singles, pairs };
}

function unionBox(
  a: [number, number, number, number],
  b: [number, number, number, number]
): [number, number, number, number] {
  return [Math.min(a[0], b[0]), Math.min(a[1], b[1]), Math.max(a[2], b[2]), Math.max(a[3], b[3])];
}

function normalizeWhitespace(s: string): string {
  return s.replace(/\s+/g, " ").trim();
}

/**
 * The hallucination guard closes one gap short of complete without this: the
 * wrap-boundary `pairs` tier below exists to recover a quote split across a
 * genuine word-wrap, but nothing stopped the model from citing a "quote"
 * that's really two adjacent-but-unrelated lines glued together — e.g.
 * quoting "Subtotal: $450.00 Sales Tax (8%): $36.00" as if it supported a
 * `total` field, when $486.00 (the value actually being claimed) never
 * appears anywhere in that quote at all. Found by evals/run.ts's held-out
 * split (an "ambiguous scan" case with no printed "Total Due" line):
 * extraction confidently computed subtotal+tax, cited the two source lines
 * as "evidence" for total, and — because that concatenation happened to
 * exist in the pairs haystack — the field came back "verified". A real
 * false clearance, not a test artifact. This check is deliberately cheap:
 * if the model's own value isn't even present in its own quote, the quote
 * cannot be real evidence for it, independent of whether it's found in the
 * document at all.
 */
function valueAppearsInQuote(value: string, quote: string): boolean {
  const strip = (s: string) => s.replace(/[,$]/g, "");
  return strip(quote).includes(strip(value));
}

function searchTier(quote: string, haystacks: Haystack[]): Haystack[] {
  const exact = haystacks.filter((h) => h.text.includes(quote));
  if (exact.length > 0) return exact;

  const lowerQuote = quote.toLowerCase();
  const caseInsensitive = haystacks.filter((h) => h.text.toLowerCase().includes(lowerQuote));
  if (caseInsensitive.length > 0) return caseInsensitive;

  const normalizedQuote = normalizeWhitespace(quote);
  return haystacks.filter((h) => normalizeWhitespace(h.text).includes(normalizedQuote));
}

/** Searches single lines first; only consults wrap-boundary pairs if nothing matched. */
function findMatches(quote: string, tiers: { singles: Haystack[]; pairs: Haystack[] }): Haystack[] {
  const singleMatches = searchTier(quote, tiers.singles);
  if (singleMatches.length > 0) return singleMatches;
  return searchTier(quote, tiers.pairs);
}

/**
 * Aligns one raw field against the real document text layer. This is the
 * hallucination guard: a value is only ever "verified" when its claimed
 * quote is independently found in the document. A value with no findable
 * quote — whether the model omitted one or the quote simply isn't there —
 * drops to "uncertain" and cannot pass a required-field control downstream.
 */
export function alignField(
  raw: RawField,
  fieldName: string,
  haystacks: { singles: Haystack[]; pairs: Haystack[] },
  normalize?: (v: string) => string
): ExtractedField<string> {
  if (raw.value == null) {
    return { field: fieldName, value: null, confidence: 0, status: "missing", evidence: [] };
  }

  if (raw.quote == null || !valueAppearsInQuote(raw.value, raw.quote)) {
    return {
      field: fieldName,
      value: raw.value,
      normalizedValue: normalize?.(raw.value),
      confidence: 0.3,
      status: "uncertain",
      evidence: [],
    };
  }

  const matches = findMatches(raw.quote, haystacks);

  if (matches.length === 0) {
    return {
      field: fieldName,
      value: raw.value,
      normalizedValue: normalize?.(raw.value),
      confidence: 0.35,
      status: "uncertain",
      evidence: [],
    };
  }

  const uniqueBoxes = new Set(matches.map((m) => m.box.join(",")));
  const status: FieldStatus = uniqueBoxes.size > 1 ? "conflicting" : "verified";

  return {
    field: fieldName,
    value: raw.value,
    normalizedValue: normalize?.(raw.value),
    confidence: status === "verified" ? 0.97 : 0.5,
    status,
    evidence: matches.slice(0, uniqueBoxes.size > 1 ? matches.length : 1).map((m) => ({
      page: m.page,
      text: raw.quote as string,
      boundingBox: m.box,
    })),
  };
}

const normalizeAlnum = (v: string) => v.toUpperCase().replace(/[^0-9A-Z]/g, "");

export function alignExtraction(raw: RawExtraction, textLayer: TextLayerLine[]): ExtractedInvoice {
  const haystacks = buildHaystacks(textLayer);
  const field = (f: RawField, name: string, normalize?: (v: string) => string) =>
    alignField(f, name, haystacks, normalize);

  const lineItems: InvoiceLineItem[] = raw.lineItems.map((li) => ({
    lineNumber: li.lineNumber,
    description: field(li.description, "description"),
    quantity: field(li.quantity, "quantity"),
    unitPrice: field(li.unitPrice, "unitPrice"),
    lineTotal: field(li.lineTotal, "lineTotal"),
  }));

  return {
    invoiceNumber: field(raw.invoiceNumber, "invoiceNumber", normalizeAlnum),
    invoiceDate: field(raw.invoiceDate, "invoiceDate"),
    dueDate: field(raw.dueDate, "dueDate"),
    supplierName: field(raw.supplierName, "supplierName"),
    supplierTaxId: field(raw.supplierTaxId, "supplierTaxId", normalizeAlnum),
    purchaseOrderNumber: field(raw.purchaseOrderNumber, "purchaseOrderNumber", normalizeAlnum),
    currency: field(raw.currency, "currency"),
    subtotal: field(raw.subtotal, "subtotal"),
    tax: field(raw.tax, "tax"),
    total: field(raw.total, "total"),
    remittanceDetails: field(raw.remittanceDetails, "remittanceDetails"),
    notes: field(raw.notes, "notes"),
    lineItems,
  };
}

File: ledgerguard/extraction/align_evidence.py
"""Aligns raw extracted fields against the real document text layer — the
hallucination guard."""

import re
from dataclasses import dataclass, field

from .pdf_text_layer import TextLayerLine


@dataclass
class Haystack:
    text: str
    page: int
    box: tuple[float, float, float, float]


@dataclass
class Evidence:
    page: int
    text: str
    bounding_box: tuple[float, float, float, float]


@dataclass
class ExtractedFieldValue:
    field: str
    value: str | None
    normalized_value: str | None = None
    confidence: float = 0.0
    status: str = "missing"  # verified | uncertain | conflicting | missing
    evidence: list[Evidence] = field(default_factory=list)

    def to_dict(self) -> dict:
        return {
            "field": self.field,
            "value": self.value,
            "normalizedValue": self.normalized_value,
            "confidence": self.confidence,
            "status": self.status,
            "evidence": [
                {"page": e.page, "text": e.text, "boundingBox": list(e.bounding_box)} for e in self.evidence
            ],
        }


_WHITESPACE = re.compile(r"\s+")


def _union_box(a: tuple[float, float, float, float], b: tuple[float, float, float, float]):
    return (min(a[0], b[0]), min(a[1], b[1]), max(a[2], b[2]), max(a[3], b[3]))


def _normalize_whitespace(s: str) -> str:
    return _WHITESPACE.sub(" ", s).strip()


def _build_haystacks(text_layer: list[TextLayerLine]) -> tuple[list[Haystack], list[Haystack]]:
    """Two search tiers: individual text-layer lines, and consecutive
    same-page pairs concatenated. Pairs exist only to recover a quote that
    fell across a PDF line-wrap boundary — they must stay a strict fallback,
    searched only when no single line matches."""
    singles = [Haystack(text=l.text, page=l.page, box=l.box) for l in text_layer]
    pairs: list[Haystack] = []
    for i in range(len(text_layer) - 1):
        a, b = text_layer[i], text_layer[i + 1]
        if a.page != b.page:
            continue
        pairs.append(Haystack(text=f"{a.text} {b.text}", page=a.page, box=_union_box(a.box, b.box)))
    return singles, pairs


def _value_appears_in_quote(value: str, quote: str) -> bool:
    """If the model's own value isn't even present in its own quote, the
    quote cannot be real evidence for it, independent of whether it's found
    in the document at all."""

    def strip(s: str) -> str:
        return s.replace(",", "").replace("$", "")

    return strip(value) in strip(quote)


def _search_tier(quote: str, haystacks: list[Haystack]) -> list[Haystack]:
    exact = [h for h in haystacks if quote in h.text]
    if exact:
        return exact
    lower_quote = quote.lower()
    case_insensitive = [h for h in haystacks if lower_quote in h.text.lower()]
    if case_insensitive:
        return case_insensitive
    normalized_quote = _normalize_whitespace(quote)
    return [h for h in haystacks if normalized_quote in _normalize_whitespace(h.text)]


def _find_matches(quote: str, singles: list[Haystack], pairs: list[Haystack]) -> list[Haystack]:
    single_matches = _search_tier(quote, singles)
    if single_matches:
        return single_matches
    return _search_tier(quote, pairs)


def align_field(
    raw: dict,
    field_name: str,
    singles: list[Haystack],
    pairs: list[Haystack],
    normalize=None,
) -> ExtractedFieldValue:
    value = raw.get("value")
    quote = raw.get("quote")

    if value is None:
        return ExtractedFieldValue(field=field_name, value=None, confidence=0, status="missing", evidence=[])

    if quote is None or not _value_appears_in_quote(value, quote):
        return ExtractedFieldValue(
            field=field_name,
            value=value,
            normalized_value=normalize(value) if normalize else None,
            confidence=0.3,
            status="uncertain",
            evidence=[],
        )

    matches = _find_matches(quote, singles, pairs)

    if not matches:
        return ExtractedFieldValue(
            field=field_name,
            value=value,
            normalized_value=normalize(value) if normalize else None,
            confidence=0.35,
            status="uncertain",
            evidence=[],
        )

    unique_boxes = {",".join(str(c) for c in m.box) for m in matches}
    status = "conflicting" if len(unique_boxes) > 1 else "verified"
    take = len(matches) if len(unique_boxes) > 1 else 1

    return ExtractedFieldValue(
        field=field_name,
        value=value,
        normalized_value=normalize(value) if normalize else None,
        confidence=0.97 if status == "verified" else 0.5,
        status=status,
        evidence=[Evidence(page=m.page, text=quote, bounding_box=m.box) for m in matches[:take]],
    )


def _normalize_alnum(v: str) -> str:
    return re.sub(r"[^0-9A-Z]", "", v.upper())


def align_extraction(raw: dict, text_layer: list[TextLayerLine]) -> dict:
    singles, pairs = _build_haystacks(text_layer)

    def f(raw_field: dict, name: str, normalize=None) -> ExtractedFieldValue:
        return align_field(raw_field, name, singles, pairs, normalize)

    line_items = []
    for li in raw["lineItems"]:
        line_items.append(
            {
                "lineNumber": li["lineNumber"],
                "description": f(li["description"], "description"),
                "quantity": f(li["quantity"], "quantity"),
                "unitPrice": f(li["unitPrice"], "unitPrice"),
                "lineTotal": f(li["lineTotal"], "lineTotal"),
            }
        )

    return {
        "invoiceNumber": f(raw["invoiceNumber"], "invoiceNumber", _normalize_alnum),
        "invoiceDate": f(raw["invoiceDate"], "invoiceDate"),
        "dueDate": f(raw["dueDate"], "dueDate"),
        "supplierName": f(raw["supplierName"], "supplierName"),
        "supplierTaxId": f(raw["supplierTaxId"], "supplierTaxId", _normalize_alnum),
        "purchaseOrderNumber": f(raw["purchaseOrderNumber"], "purchaseOrderNumber", _normalize_alnum),
        "currency": f(raw["currency"], "currency"),
        "subtotal": f(raw["subtotal"], "subtotal"),
        "tax": f(raw["tax"], "tax"),
        "total": f(raw["total"], "total"),
        "remittanceDetails": f(raw["remittanceDetails"], "remittanceDetails"),
        "notes": f(raw["notes"], "notes"),
        "lineItems": line_items,
    }


def extracted_to_dict(extracted: dict) -> dict:
    """Converts an align_extraction() result (ExtractedFieldValue objects,
    the shape every matching/*.py function expects via attribute access)
    into a plain JSON-safe dict — the shape the `invoices.extracted` column
    and templates expect, matching fixtures/scenarios.py's static shape."""

    def convert(v):
        return v.to_dict() if isinstance(v, ExtractedFieldValue) else v

    return {
        **{k: convert(v) for k, v in extracted.items() if k != "lineItems"},
        "lineItems": [{k: convert(v) for k, v in li.items()} for li in extracted["lineItems"]],
    }
